File: src/shared/api/cache.ts
import { invoke } from '@tauri-apps/api/core';

export const ipcInflight = new Map<string, Promise<unknown>>();

export function cachedInvoke<T>(key: string, fn: () => Promise<T>): Promise<T> {
  const existing = ipcInflight.get(key);
  if (existing) return existing as Promise<T>;
  const promise = fn().finally(() => ipcInflight.delete(key));
  ipcInflight.set(key, promise);
  return promise;
}

export function invalidateCache(_keys?: string[]) {
  if (!_keys || _keys.length === 0) {
    ipcInflight.clear();
    return;
  }

  for (const prefix of _keys) {
    for (const key of [...ipcInflight.keys()]) {
      if (key.startsWith(prefix) || key.includes(prefix)) {
        ipcInflight.delete(key);
      }
    }
  }
}

// ── Persistent cache (second layer, backed by disk) ──

const DEFAULT_TTL = 300; // 5 minutes

/**
 * Get a value from the persistent (disk-backed) cache.
 * Returns null if the key is not found or has expired.
 */
export async function persistentCacheGet(key: string): Promise<string | null> {
  const result = await invoke<string | null>('cache_get', { key });
  return result;
}

/**
 * Set a value in the persistent (disk-backed) cache.
 * @param key Cache key
 * @param value Value to store (will be stored as string)
 * @param ttlSecs Time-to-live in seconds (default: 300 / 5 minutes)
 */
export async function persistentCacheSet(key: string, value: string, ttlSecs: number = DEFAULT_TTL): Promise<void> {
  await invoke('cache_set', { key, value, ttlSecs });
}

/**
 * Invalidate all cache entries whose key starts with the given prefix.
 */
export async function persistentCacheInvalidate(keyPrefix: string): Promise<void> {
  await invoke('cache_invalidate', { keyPrefix });
}

/**
 * Evict all expired entries from the persistent cache.
 * Returns the number of entries removed.
 */
export async function persistentCacheEvictExpired(): Promise<number> {
  return await invoke<number>('cache_evict_expired');
}

/**
 * Two-layer cached invoke: checks in-memory inflight cache first,
 * then persistent disk cache, and finally calls the function.
 */
export async function twoLayerCachedInvoke<T>(
  key: string,
  fn: () => Promise<T>,
  ttlSecs: number = DEFAULT_TTL,
): Promise<T> {
  // Layer 1: in-memory inflight dedup
  const existing = ipcInflight.get(key);
  if (existing) return existing as Promise<T>;

  const promise = (async () => {
    // Layer 2: persistent disk cache
    try {
      const cached = await persistentCacheGet(key);
      if (cached !== null) {
        return JSON.parse(cached) as T;
      }
    } catch {
      // persistent cache miss or parse error — fall through
    }

    // Layer 3: actual invoke
    const result = await fn();

    // Store in persistent cache in background (don't block return)
    try {
      await persistentCacheSet(key, JSON.stringify(result), ttlSecs);
    } catch {
      // persistent cache write failure — non-critical
    }

    return result;
  })().finally(() => ipcInflight.delete(key));

  ipcInflight.set(key, promise);
  return promise;
}

File: src-tauri/src/commands/cache.rs
use parking_lot::Mutex;
use std::sync::Arc;

use crate::error::LauncherError;
use crate::persistent_cache::PersistentCache;

pub struct PersistentCacheState(pub Arc<Mutex<PersistentCache>>);

#[tauri::command]
pub async fn cache_get(key: String, state: tauri::State<'_, PersistentCacheState>) -> Result<Option<String>, LauncherError> {
    let mut cache = state.0.lock();
    Ok(cache.get(&key))
}

#[tauri::command]
pub async fn cache_set(key: String, value: String, ttl_secs: u64, state: tauri::State<'_, PersistentCacheState>) -> Result<(), LauncherError> {
    let mut cache = state.0.lock();
    cache.set(&key, &value, ttl_secs, None);
    cache.save()?;
    Ok(())
}

#[tauri::command]
pub async fn cache_invalidate(key_prefix: String, state: tauri::State<'_, PersistentCacheState>) -> Result<(), LauncherError> {
    let mut cache = state.0.lock();
    cache.invalidate(&key_prefix);
    cache.save()?;
    Ok(())
}

#[tauri::command]
pub async fn cache_evict_expired(state: tauri::State<'_, PersistentCacheState>) -> Result<usize, LauncherError> {
    let mut cache = state.0.lock();
    let removed = cache.evict_expired();
    cache.save()?;
    Ok(removed)
}
